sdk/keyvault/azure-security-keyvault-keys/src/cryptography/cryptography_client.cpp

```diff
@@ -0,0 +1,389 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// SPDX-License-Identifier: MIT
+
+#include <azure/core/credentials/credentials.hpp>
+#include <azure/core/cryptography/hash.hpp>
+#include <azure/core/exception.hpp>
+#include <azure/core/http/http.hpp>
+#include <azure/core/http/policies/policy.hpp>
+
+#include "azure/keyvault/keys/cryptography/cryptography_client.hpp"
+#include "azure/keyvault/keys/cryptography/local_cryptography_provider_factory.hpp"
+#include "azure/keyvault/keys/key_operation.hpp"
+
+#include <memory>
+#include <string>
+#include <vector>
+
+using namespace Azure::Security::KeyVault::Keys::Cryptography;
+using namespace Azure::Security::KeyVault::Keys::Cryptography::_detail;
+using namespace Azure::Core::Http;
+using namespace Azure::Core::Http::Policies;
+using namespace Azure::Core::Http::Policies::_internal;
+
+namespace {
+// 1Mb at a time
+const size_t DefaultStreamDigestReadSize = 1024 * 1024;
+
+inline std::vector<uint8_t> CreateDigest(
+    SignatureAlgorithm algorithm,
+    Azure::Core::IO::BodyStream& data)
+{
+  // Use heap for the reading buffer.
+  auto heapBuffer = std::make_unique<std::vector<uint8_t>>(DefaultStreamDigestReadSize);
+  auto* buffer = heapBuffer.get()->data();
+  auto hasAlgo = algorithm.GetHashAlgorithm();
+  for (uint64_t read = data.Read(buffer, DefaultStreamDigestReadSize); read > 0;
+       read = data.Read(buffer, DefaultStreamDigestReadSize))
+  {
+    hasAlgo->Append(buffer, static_cast<size_t>(read));
+  }
+  return hasAlgo->Final();
+}
+
+inline std::vector<uint8_t> CreateDigest(
+    SignatureAlgorithm algorithm,
+    std::vector<uint8_t> const& data)
+{
+  auto hasAlgo = algorithm.GetHashAlgorithm();
+  return hasAlgo->Final(data.data(), data.size());
+}
+} // namespace
+
+void CryptographyClient::Initialize(std::string const&, Azure::Core::Context const& context)
+{
+  if (m_provider != nullptr)
+  {
+    return;
+  }
```

Comment:
I will try to add tests to see what happen in multi-thread and if that' something we support in other languages for a crypto client.   I would suspect that a better approach would be to share a key client on multiple threads and create a crypto client from it on each thread instead of sharing the same crypto client in multiple threads.